flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/common/WindowStartEndPropertiesRule.scala

```
@@ -54,34 +56,75 @@ class WindowStartEndPropertiesRule
     transformed.push(LogicalWindowAggregate.create(
-      agg.getWindow,
-      Seq(
-        NamedWindowProperty("w$start", WindowStart(agg.getWindow.aliasAttribute)),
-        NamedWindowProperty("w$end", WindowEnd(agg.getWindow.aliasAttribute))
-      ), agg)
+      window,
+      properties,
+      agg)
     )
 
     // forward window start and end properties
     transformed.project(
-      innerProject.getProjects ++ Seq(transformed.field("w$start"), transformed.field("w$end")))
+      innerProject.getProjects ++ properties.map(np => transformed.field(np.name)))
 
     def replaceGroupAuxiliaries(node: RexNode): RexNode = {
       node match {
         case c: RexCall if WindowStartEndPropertiesRule.isWindowStart(c) =>
           // replace expression by access to window start
           rexBuilder.makeCast(c.getType, transformed.field("w$start"), false)
+
         case c: RexCall if WindowStartEndPropertiesRule.isWindowEnd(c) =>
           // replace expression by access to window end
           rexBuilder.makeCast(c.getType, transformed.field("w$end"), false)
+
+        case c: RexCall if WindowStartEndPropertiesRule.isWindowRowtime(c) =>
+          if (isProctime) {
+            throw ValidationException("A proctime window cannot provide a rowtime attribute.")
+          } else if (isRowtime) {
+            // replace expression by access to window rowtime
+            transformed.field("w$rowtime")
+          } else {
+            throw TableException("Accessing the rowtime attribute of a window is not yet " +
+              "supported in a batch environment.")
+          }
+
+        case c: RexCall if WindowStartEndPropertiesRule.isWindowProctime(c) =>
+          if (isProctime) {
+            // replace expression by access to window proctime
+            transformed.field("w$proctime")
+          } else {
+            throw ValidationException("Proctime is not supported in a batch environment.")
+          }
```

We can throw this exception in a Stream rowtime window if we want query `TUMBLE_PROCTIME`, So I thinks this message should be improve or add a `isRowtime` process.